feat: Custom Fields OR Add-Ons with a fee associated

asadath1395 opened the pull request:
## What does this PR do?

Custom Fields OR Add-Ons with a fee associated

- Fixes #7075
- Fixes CAL-3967

/claim #7075

## Visual Demo (For contributors especially)

#### Video Demo (if applicable):

[https://www.loom.com/share/c4fb241826874cc894837e85534cfd7d?sid=389aae54-dc26-428d-a902-6f4ae72bc25e](https://www.loom.com/share/c4fb241826874cc894837e85534cfd7d?sid=389aae54-dc26-428d-a902-6f4ae72bc25e)

**Right now pricing is supported for Number, Select, MultiSelect, Checkbox, Checkbox group and Radio group input type**

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

Check the Video Demo

asadath1395 committed:
feat: Addons to bookings

asadath1395 committed:
Add tests

asadath1395 committed:
Add tests

asadath1395 commented on packages/features/bookings/Booker/components/BookEventForm/BookingFields.tsx:
```diff
@@ -134,8 +145,50 @@ export const BookingFields = ({
+              ...field,
+              label,
+              ...(fieldsThatSupportLabelAsSafeHtml.includes(field.type) && field.labelAsSafeHtml
+                ? { labelAsSafeHtml: markdownToSafeHTML(label) }
```
Should have avoided `markdownToSafeHTML` as it says its an expensive function, but had to re-call it just to append the price

retrogtx replied:
instead of this, can we not use markdown-it (already used in our codebase at `apps/web/lib/video/[uid]/getServerSideProps.ts`)

asadath1395 replied:
`markdownToSafeHTML` is internally using `markdown-it`. Ref: https://github.com/calcom/cal.com/blob/9838488068ae0bf3a001431be7788f046cc81485/packages/lib/markdownToSafeHTML.ts#L3
https://github.com/calcom/cal.com/blob/9838488068ae0bf3a001431be7788f046cc81485/packages/lib/markdownIt.ts#L3

asadath1395 committed:
Fix price field not shown immediately after enabling payment for an event

retrogtx reviewed (commented):
hey some type checks are failing, good pr nonetheless

asadath1395 committed:
Add support for all currencies in addons and fix type errors

asadath1395 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feature/addons

asadath1395 commented:
> hey some type checks are failing, good pr nonetheless

@retrogtx Types should be fixed now

zomars committed:
Merge branch 'main' into feature/addons

retrogtx committed:
Merge branch 'main' into feature/addons

retrogtx committed:
Merge branch 'main' into feature/addons

asadath1395 commented:
@keithwillcode Can we get this merged? Thanks

asadath1395 committed:
Merge branch 'main' into feature/addons

asadath1395 committed:
Merge branch 'main' into feature/addons

asadath1395 committed:
Merge branch 'main' into feature/addons

asadath1395 commented:
@keithwillcode Anything to be done here to get this reviewed?

retrogtx commented:
@asadath1395 it will be reviewed soon

asadath1395 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feature/addons

asadath1395 committed:
Merge branch 'main' into feature/addons

asadath1395 committed:
Merge branch 'main' into feature/addons

asadath1395 commented:
@retrogtx 
![mr-bean-waiting](https://github.com/user-attachments/assets/eb5abff4-f3b0-48bd-bb2c-af047dbc6f43)

retrogtx commented:
I understand, this needs a review from the foundation which usually takes time as they are occupied with other stuff :)

hbjORbj commented on packages/lib/payment/handlePayment.ts:
```diff
@@ -66,25 +71,80 @@ const handlePayment = async ({
+              break;
+            case "boolean":
+              if (evt.responses[field.name]?.value) {
+                addonsPrice += typedInput.price || 0;
+              }
+              break;
+            case "select":
+            case "radio":
+              // For select and radio, find the selected option and add its price
+              const selectedValue = evt.responses[field.name]?.value;
+              let selectedOption = typedInput.options?.find((opt) => opt.value === selectedValue);
+              // For radio, the value coming is the label itself
+              if (field.type === "radio") {
+                selectedOption = typedInput.options?.find((opt) => {
+                  const formattedValue = `${opt.value} (${Intl.NumberFormat("en", {
+                    style: "currency",
+                    currency: paymentCurrency,
+                  }).format(opt.price || 0)})`;
+                  return formattedValue === selectedValue;
+                });
+              }
+              addonsPrice += selectedOption?.price || 0;
+              break;
+            case "checkbox":
+            case "multiselect":
+              // For checkbox and multiselect, add prices of all selected options
+              const response = evt.responses[field.name]?.value;
+              const selectedValues = Array.isArray(response) ? response : response ? [response] : [];
+              selectedValues.forEach((value) => {
+                const option = typedInput.options?.find((opt) => opt.value === value);
+                addonsPrice += option?.price || 0;
+              });
+              break;
+          }
+        }
+      }
+    });
+
+    totalAmount += convertToSmallestCurrencyUnit(addonsPrice, paymentCurrency);
+  }
```
Not very readable. I asked AI to refactor this, and here's what I got. Can you use this code and test? If not working, feel free to tweak it
```tsx
if ((bookingFields || [])?.length > 0) {
  let addonsPrice = 0;
  
  // Process each booking field
  (bookingFields || []).forEach((field) => {
    // Skip fields that don't support pricing
    if (!fieldTypesConfigMap[field.type]?.supportsPricing && 
        !fieldTypesConfigMap[field.type]?.optionsSupportPricing) {
      return;
    }
    
    // Type assertion for price-related fields
    const typedInput = field as { price?: number; options?: { value: string; price?: number }[] };
    const response = evt.responses?.[field.name];
    
    // Skip if no response for this field
    if (!response) return;
    
    switch (field.type) {
      case "number":
        // Multiply the numeric value by the field's price
        addonsPrice += Number(response.value || 0) * (typedInput.price || 0);
        break;
        
      case "boolean":
        // Add price if boolean field is true
        if (response.value) {
          addonsPrice += typedInput.price || 0;
        }
        break;
        
      case "select":
      case "radio":
        // For select and radio, find the selected option and add its price
        const selectedValue = response.value;
        let selectedOption;
        
        if (field.type === "radio") {
          // For radio, the value coming is the label itself
          selectedOption = typedInput.options?.find((opt) => {
            const formattedValue = `${opt.value} (${Intl.NumberFormat("en", {
              style: "currency",
              currency: paymentCurrency,
            }).format(opt.price || 0)})`;
            return formattedValue === selectedValue;
          });
        } else {
          // For select, match by direct value
          selectedOption = typedInput.options?.find((opt) => opt.value === selectedValue);
        }
        
        addonsPrice += selectedOption?.price || 0;
        break;
        
      case "checkbox":
      case "multiselect":
        // For checkbox and multiselect, add prices of all selected options
        const responseValue = response.value;
        const selectedValues = Array.isArray(responseValue) 
          ? responseValue 
          : responseValue 
            ? [responseValue] 
            : [];
            
        selectedValues.forEach((value) => {
          const option = typedInput.options?.find((opt) => opt.value === value);
          addonsPrice += option?.price || 0;
        });
        break;
    }
  });

  totalAmount += convertToSmallestCurrencyUnit(addonsPrice, paymentCurrency);
}
```

hbjORbj commented on packages/features/bookings/Booker/components/BookEventForm/BookEventForm.tsx:
```diff
@@ -84,6 +84,11 @@ export const BookEventForm = ({
 
+  const paymentCurrency = useMemo(() => {
+    if (!eventType) return "USD";
+    return getPaymentAppData(eventType)?.currency || "USD";
```
Is this `currency` type string? It makes more sense for `currency` to be an enum

asadath1395 replied:
Yes it is string, using already existing function code from https://github.com/calcom/cal.com/blob/880fe151c079b38ce12d8828fcb8537004a22df8/packages/lib/getPaymentAppData.ts#L37

asadath1395 committed:
Merge branch 'feature/addons' of https://github.com/asadath1395/cal.com into feature/addons

asadath1395 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feature/addons

asadath1395 committed:
Address review comments

asadath1395 commented on packages/app-store/_utils/payments/currencyConversions.ts:
```diff
@@ -32,3 +32,14 @@ export const convertFromSmallestToPresentableCurrencyUnit = (amount: number, cur
+    const parts = formatter.formatToParts(1);
+    const currencyPart = parts.find((part) => part.type === "currency");
+    return currencyPart?.value || "$";
+  } catch {
```
Added

asadath1395 commented on packages/app-store/_utils/payments/currencyConversions.ts:
```diff
@@ -32,3 +32,14 @@ export const convertFromSmallestToPresentableCurrencyUnit = (amount: number, cur
+export const getCurrencySymbol = (currencyCode: string): string => {
+  try {
+    const formatter = new Intl.NumberFormat("en", { style: "currency", currency: currencyCode });
+    const parts = formatter.formatToParts(1);
```
Added comment

asadath1395 commented on packages/app-store/_utils/payments/currencyConversions.ts:
```diff
@@ -32,3 +32,14 @@ export const convertFromSmallestToPresentableCurrencyUnit = (amount: number, cur
   return amount / 100;
 };
+
+export const getCurrencySymbol = (currencyCode: string): string => {
```
As i mentioned in above comment, we are getting the value from already existing function. We can change that in followup PR as it could have greater impact

asadath1395 commented on packages/features/bookings/Booker/components/BookEventForm/BookingFields.tsx:
```diff
@@ -134,8 +147,50 @@ export const BookingFields = ({
+        // Add price display for custom inputs with prices
+        let fieldWithPrice = field;
+
+        if (isPaidEvent) {
```
Made it more readable by splitting it up into functions. Can you check now?

asadath1395 committed:
Merge branch 'main' into feature/addons

asadath1395 commented:
@hbjORbj Thanks for reviewing this PR, appreciate you taking the time. I have made the changes as you suggested, could you review this once more.

asadath1395 committed:
Add tests to cover all input types

asadath1395 committed:
Merge branch 'main' into feature/addons

retrogtx commented:
There's some good suggestions from mrge

You should look into it, and close the ones that don't apply!

Kinbaum commented:
@hbjORbj any update on this when this PR will get merged?

asadath1395 committed:
Merge with upstream

asadath1395 committed:
Merge branch 'feature/addons' of https://github.com/asadath1395/cal.com into feature/addons

asadath1395 committed:
Address cubic-dev-ai comments

asadath1395 committed:
Merge branch 'main' into feature/addons

kart1ka commented:
Hey @asadath1395, pls address the cubic comments.

asadath1395 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feature/addons

asadath1395 committed:
Merge branch 'feature/addons' of https://github.com/asadath1395/cal.com into feature/addons

asadath1395 committed:
Remove openapi.json from git tracking

asadath1395 committed:
Revert changes

asadath1395 committed:
Revert unrelated linter/formatter changes

asadath1395 committed:
Address cubic-dev-ai comments

asadath1395 commented on packages/features/eventtypes/components/tabs/advanced/EventAdvancedTab.tsx:
```diff
@@ -552,6 +553,21 @@
     userEmail = removePlatformClientIdFromEmail(userEmail, platformContext.clientId);
   }
 
+  const formValues = formMethods.watch();
```
Done!

asadath1395 commented on packages/features/eventtypes/components/tabs/advanced/EventAdvancedTab.tsx:
```diff
@@ -552,6 +553,21 @@ export const EventAdvancedTab = ({
+  }, [formValues.metadata]);
+
+  const isPaidEvent = useMemo(
+    () => !Number.isNaN(paymentAppData.price) && paymentAppData.price > 0,
```
We are only concerned when users have enabled payments apps and setup base pricing

asadath1395 commented on packages/features/bookings/Booker/components/BookEventForm/BookingFields.tsx:
```diff
@@ -6,30 +6,81 @@
+    }).format(price)})`;
+
+  const getFieldWithDirectPricing = (field: Fields[number]) => {
+    if (!fieldTypesConfigMap[field.type]?.supportsPricing || !field.label || !field.price) {
```
We only want to show when user's have entered a value > 0

asadath1395 commented on packages/features/bookings/Booker/components/BookEventForm/BookingFields.tsx:
```diff
@@ -6,30 +6,81 @@
+        const optionPrice = option.price;
+
+        // Only add price to label if there's a price
+        if (!optionPrice) return option;
```
We only want to show when user's have entered a value > 0

asadath1395 commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -428,26 +439,52 @@
+                      // Right now we use label of the option as the value of the option. It allows use to not separately lookup the optionId to know the optionValue
+                      // It has the same drawback that if the label is changed, the value of the option will change. It is not a big deal for now.
+                      const newOptions = [...(value || [])];
+                      newOptions.splice(index, 1, {
```
This is expected since we are creating a new option, so don't want pricing to carry over

asadath1395 committed:
Addres cubic-dev-ai comment

asadath1395 commented on packages/features/form-builder/schema.ts:
```diff
@@ -82,7 +82,15 @@ const baseFieldSchema = z.object({
+      z.object({
+        label: z.string(),
+        value: z.string(),
+        price: z.coerce.number().min(0).optional(),
```
`min(0)` will take care of all non-negative numbers

asadath1395 commented on packages/features/form-builder/testUtils.ts:
```diff
@@ -158,6 +161,53 @@
+      fireEvent.click(dialog.getByText("Add an Option"));
+    },
+    fillInOption: async ({
+      dialog,
```
Fixed!

asadath1395 commented on packages/features/form-builder/schema.ts:
```diff
@@ -82,7 +82,15 @@ const baseFieldSchema = z.object({
+      z.object({
+        label: z.string(),
+        value: z.string(),
+        price: z.coerce.number().min(0).optional(),
```
Only numbers are going to come since we are only allowing users to enter numbers

asadath1395 commented on packages/features/bookings/Booker/components/BookEventForm/BookingFields.tsx:
```diff
@@ -6,30 +6,81 @@
+      currency: paymentCurrency,
+    }).format(price)})`;
+
+  const getFieldWithDirectPricing = (field: Fields[number]) => {
```
Only numbers are going to come since we are only allowing users to enter numbers

asadath1395 commented on packages/lib/payment/handlePayment.ts:
```diff
@@ -66,25 +71,104 @@ const handlePayment = async ({
   const paymentOption = apps?.[paymentAppCredentials.appId].paymentOption || "ON_BOOKING";
+  const paymentCurrency = apps?.[paymentAppCredentials.appId].currency;
+
+  let totalAmount = apps?.[paymentAppCredentials.appId].price || 0;
```
The base price is stored as the smallest currency in the DB

asadath1395 commented on packages/lib/payment/handlePayment.ts:
```diff
@@ -66,25 +71,104 @@
+      switch (field.type) {
+        case "number":
+          // Multiply the numeric value by the field's price
+          addonsPrice += Number(response.value || 0) * (typedInput.price || 0);
```
Fixed!

asadath1395 commented on packages/features/eventtypes/components/tabs/advanced/EventAdvancedTab.tsx:
```diff
@@ -580,6 +596,8 @@ export const EventAdvancedTab = ({
             return field.name === "location" ? true : field.required;
           }}
+          showPriceField={isPaidEvent}
+          paymentCurrency={paymentAppData?.currency || "USD"}
```
Fixed!

asadath1395 commented:
@kart1ka All cubic comments are addressed, please review.
@hbjORbj Could you please take sometime to review this PR? I have addressed all your comments. I would appreciate if we could get this merged.

Thanks!

asadath1395 commented:
@hbjORbj Could you take some of your precious time this week to review this PR? Thanks

asadath1395 committed:
Merge branch 'main' into feature/addons

asadath1395 committed:
Update to use test id instead of hardcoded label

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

asadath1395 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feature/addons

asadath1395 committed:
Address coderabbits review comments

asadath1395 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feature/addons

asadath1395 committed:
Fix failing unit test

kart1ka committed:
Merge branch 'main' into feature/addons

asadath1395 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feature/addons

anikdhabal committed:
Merge branch 'main' into feature/addons

Devanshusharma2005 committed:
Merge branch 'main' into feature/addons

Devanshusharma2005 committed:
some fixes...

hbjORbj committed:
Merge branch 'main' into feature/addons

Devanshusharma2005 reviewed (commented):
Could you please resolve the failing tests.

asadath1395 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feature/addons

asadath1395 committed:
Fix tests

asadath1395 committed:
Merge branch 'main' into feature/addons

Devanshusharma2005 committed:
Merge branch 'main' into feature/addons

asadath1395 committed:
Merge branch 'main' into feature/addons

Devanshusharma2005 commented:
still the e2e test failing.

asadath1395 committed:
Merge branch 'main' into feature/addons

asadath1395 committed:
Merge branch 'main' into feature/addons

Devanshusharma2005 committed:
Merge branch 'main' into feature/addons

Devanshusharma2005 reviewed (commented):
There is still a merge conflict. Could you please address it.

asadath1395 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feature/addons

asadath1395 committed:
Merge branch 'main' into feature/addons

asadath1395 commented:
> There is still a merge conflict. Could you please address it.

Fixed!

Devanshusharma2005 committed:
Merge branch 'main' into feature/addons

asadath1395 committed:
Fix import

asadath1395 committed:
Merge branch 'feature/addons' of https://github.com/asadath1395/cal.com into feature/addons

asadath1395 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feature/addons

PeerRich committed:
Merge branch 'main' into feature/addons

asadath1395 committed:
Move handlePayment test file to appropriate location

asadath1395 committed:
Fix test

volnei committed:
Merge branch 'main' into feature/addons

volnei reviewed (approved):
LGTM. Thank you for this awesome contribution! ❤
